src/Features/Core/Portable/SignatureHelp/SignatureHelpServiceWithProviders.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Shared.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SignatureHelp;
+
+/// <summary>
+/// Base class for <see cref="SignatureHelpService"/>'s that delegate to <see cref="ISignatureHelpProvider"/>'s.
+/// </summary>
+internal abstract class SignatureHelpServiceWithProviders : SignatureHelpService
+{
+    private readonly LanguageServices _services;
+    private ImmutableArray<ISignatureHelpProvider> _providers;
+
+    protected SignatureHelpServiceWithProviders(LanguageServices services)
+    {
+        _services = services;
```

Comment:
these services are not used except to get the Language.  We can make this just take in a `string language` instead.  And have the C# and VB versions not use factories, and instead just pass in their language name directly.